### PR TITLE
Provide a mechanism for pinning type signatures formats during printing

### PR DESCRIPTION
**Pull request by baronfel**
In FSAC we have some pretty convoluted logic for undoing the compiler's default printing of generic type parameters for types that use the suffix form (ie `int list`).

We'd love a way to explicit ask NicePrint to render the type signature in a given way, either prefix or suffix, regardless of the derived flag on the type itself.

To that end, this PR:

* introduces a flag on DisplayEnv that represents the rendering mode, defaulting to delegating to the flag on the type,
* provides a way for an FCS consumer to configure the display environment to prefix or suffix mode, and
* uses the flag to potentially overwrite the prefix/suffix logic when a type with generic parameters is printed.

### CODE REVIEW

File: src/fsharp/TypedTreeOps.fs [hunk 1]
```diff
@@ -2737,6 +2737,11 @@ module SimplifyTypes =
 // Print Signatures/Types
 //-------------------------------------------------------------------------- 
 
+type GenericParameterStyle =
+| Implicit
+| Prefix
+| Suffix
```
Comment (cartermp) [inline src/fsharp/TypedTreeOps.fs:2743]:
```suggestion
    | Implicit
    | Prefix
    | Suffix
```

File: src/fsharp/TypedTreeOps.fsi [hunk 2]
```diff
@@ -960,6 +960,15 @@ module PrettyTypes =
...
     val PrettifyInstAndCurriedSig : TcGlobals -> TyparInst * TTypes * CurriedArgInfos * TType -> (TyparInst * TTypes * CurriedArgInfos * TType) * TyparConstraintsWithTypars
 
+/// Describes how generic type parameters in a type will be formatted during printing
+type GenericParameterStyle =
+/// Use the IsPrefixDisplay member of the TyCon to determine the style
+| Implicit
+/// Force the prefix style: List<int>
+| Prefix
+/// Force the suffix style: int List
+| Suffix
```
Comment (cartermp) [inline src/fsharp/TypedTreeOps.fsi:970]:
```suggestion
    /// Use the IsPrefixDisplay member of the TyCon to determine the style
    | Implicit
    /// Force the prefix style: List<int>
    | Prefix
    /// Force the suffix style: int List
    | Suffix
```

File: src/fsharp/TypedTreeOps.fsi [hunk 3]
```diff
@@ -986,7 +995,8 @@ type DisplayEnv =
       printVerboseSignatures : bool
       g: TcGlobals
       contextAccessibility: Accessibility
-      generatedValueLayout:(Val -> layout option) }
+      generatedValueLayout:(Val -> layout option)
```
Comment (cartermp) [inline src/fsharp/TypedTreeOps.fsi:998]:
```suggestion
      generatedValueLayout: (Val -> layout option)
```

File: src/fsharp/TypedTreeOps.fs [hunk 4]
```diff
@@ -2796,7 +2802,8 @@ type DisplayEnv =
         printVerboseSignatures = false
         g = tcGlobals
         contextAccessibility = taccessPublic
-        generatedValueLayout = (fun _ -> None) }
+        generatedValueLayout = (fun _ -> None)
+        genericParameterStyle = GenericParameterStyle.Implicit }
```
Comment (cartermp) [inline src/fsharp/TypedTreeOps.fs:2806]: Is this the correct default?

**Review by cartermp — COMMENTED** [review]
Formatting changes. Note that NicePrint is also used to print signatures. So some tests might fail because of that and the sig printing would have to be adjusted to match the current behavior.

File: src/fsharp/TypedTreeOps.fs (hunk 4 above)
Comment (baronfel) [reply src/fsharp/TypedTreeOps.fs:2806]: This preserves the behavior currently, which is to print based on the state of the flag on the TyCon.

I could see a future FCS release where we say 'nah let's just go for Prefix by default' and we'd change it here, though.

**Comment by baronfel** [comment]
I would very much be surprised if any tests failed, as I intended to make this feature opt-in entirely.

**Review by cartermp — APPROVED** [review]

**Review by KevinRansom — APPROVED** [review]
Nice, thank you.

**Pull request merged**